### ISSUE TITLE
Cannot configure bind address for http/https gateway server

### ISSUE THREAD

**Issue by bploetz** [ExpressGateway/express-gateway#843]
(Note: I reported this on gitter and @kevinswiber suggested I file a bug.)

<img width="683" alt="screen shot 2018-11-13 at 7 11 05 pm" src="https://user-images.githubusercontent.com/92764/48451324-f574d800-e777-11e8-9bf7-adc07b64b5df.png">

There's a way to specify the bind address for the admin http server (via admin->hostname in gateway.config.yml) but you can't currently configure the bind address for the main gateway http/https servers. This is specifically problematic when trying to run in a Docker container where you need it to bind to all interfaces (0.0.0.0) when running in "bridge" network mode, for example.

I took a stab at filing a PR for this. While adding support for the new config value is trivial, the problem I ran into is maintaining backwards compatibility such that folks that currently don't have the hostname defined get the same behavior (binds to '::' for IPv6, etc), but when I *do* specify the hostname, the tests break with errors like:

```
  5) exact host name configuration
       should not serve sub.abc.acme.com :
     TypeError: Cannot read property 'address' of undefined
      at Test.serverAddress (node_modules/supertest/lib/test.js:55:18)
      at new Test (node_modules/supertest/lib/test.js:36:12)
      at Object.obj.(anonymous function) [as get] (node_modules/supertest/index.js:25:14)
      at prepareScenario (test/common/routing.helper.js:16:35)
      at Context.<anonymous> (test/common/routing.helper.js:73:30)

```

I *think* this is because node's `net.server.address()` returns an object if you don't specify a hostname (current behavior) or a String if you do (new behavior). (See https://nodejs.org/api/net.html#net_server_address), but I can't tell if this is just an issue with the tests, or if the rabbit hole goes deeper.

This may be easier than I think and I'm just not familiar with the code base. Any pointers in the right direction so I can file a PR would be much appreciated. Thanks!

**Comment by XVincentX** [comment]
@bploetz Hey,

this is great — thanks a lot for filing the issue and trying to look into the code base. Do you have any specific branch I can look at and see your code? Probably you're this close to the solution :)

**Comment by bploetz** [comment]
@XVincentX yup, will push that up when I get to work in a couple hours. Stay tuned....

**Comment by bploetz** [comment]
@XVincentX See https://github.com/bploetz/express-gateway/commit/352cf7ebbda187959f7fa6a46e07d0cfe1d958c3

If you uncomment lines 23 and/or 46 in lib/gateway/index.js and run the tests you'll see what I mean about the failures noted above. Note that I did not force set the hostname to 'localhost' like we do for the admin server so that it is backwards compatible with how it behaves today.

Any questions let me know. Thanks!

**Comment by bploetz** [comment]
@XVincentX I meant to mention that at runtime the gateway seems to function just fine when I configure a hostname to bind to with these code changes, so may just be a test issue?

```
const runningApp = httpServer.listen(config.gatewayConfig.http.port, config.gatewayConfig.http.hostname, () => {
```

Server:
```
2018-11-15T14:02:30.004Z [EG:gateway] info: gateway http server listening on 0.0.0.0:8080
2018-11-15T14:02:30.004Z [EG:admin] info: admin http server listening on 0.0.0.0:9876
```

And requests are processed just fine.

**Comment by XVincentX** [comment]
Got it. I'll try to have a look today or tomorrow. Thanks!

**Comment by XVincentX** [comment]
@bploetz I'm sorry I haven't been able to look into this yet — I've been traveling the whole week but I'm finally settled back home. I hope to give you some updates later today.

**Comment by bploetz** [comment]
@XVincentX no need to apologize, but thanks for the update. Let me know if I can help at all.

**Comment by XVincentX** [comment]
@bploetz I've created #845 with your changes and added couple of comments. I think the changes are almost legit, we just need to add some tests and polish the code a little bit.

See you in the pull request!

**Comment by XVincentX** [comment]
This has been closed by #846 .
